### ISSUE TITLE
AutoGraph unexpected indent in tf-nightly-gpu-2.1.0.dev20191103

### ISSUE THREAD

**Issue by netw0rkf10w** [tensorflow/tensorflow#33962]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): `tf-nightly-gpu-2.1.0.dev20191103`
- Python version: 3.5.2
- CUDA/cuDNN version: 10.0/7.1
- GPU model and memory: GTX 1080 Ti

**Describe the current behavior**

After upgrading to `tf-nightly-gpu-2.1.0.dev20191103` from `tensorflow-gpu-2.0.0`, I obtained this error when running my code:

```
WARNING:tensorflow:AutoGraph could not transform <bound method CRFLayer.mean_field of <models.crf_layer.CRFLayer object at 0x7f6124237b00>> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: unexpected indent (<unknown>, line 36)
```
I did `export AUTOGRAPH_VERBOSITY=10` but did not observe any other types of output other than the above. It says **unexpected indent** so I guess something changed in the parsing of Python code when building the graph? The problem does not occur on `tensorflow-gpu-2.0.0` (I installed `tf-nightly-gpu-2.1.0.dev20191103` because I need to be able to `load_weights(pretrained_weights, by_name=True, skip_mismatch=True)`, which is not available in 2.0.0. There is another bug with this function that I will report in a separate issue.)

**Describe the expected behavior**
Like in TF 2.0.0: no AutoGraph warning.

**Code to reproduce the issue**
Unfortunately my code has a lot of dependencies and I was unable to create a minimal reproducible example. But from the warning message I guess it's easy enough to check in the source code.

**Comment by oanush** [comment]
@netw0rkf10w ,
Will you be able to provide Colab link with code you are trying to execute ?Thanks!

**Comment by netw0rkf10w** [comment]
Hi @oanush. Unfortunately this code has a custom TF op in C++ that needs to be compiled. I'm not sure it's possible to do that in Colab. Let me know if sharing a zip file of source code is ok.

**Comment by gowthamkpr** [comment]
@netw0rkf10w Please share the zip file of source code. Thanks!

**Comment by gowthamkpr** [comment]
@netw0rkf10w Can you please share the source code?

**Comment by netw0rkf10w** [comment]
@gowthamkpr Sorry for the delay, I have been overwhelmed over the last days. I'm preparing the code.

**Comment by gowthamkpr** [comment]
Closing this issue as it has been inactive for more than 2 weeks. Please add additional comments and we can open this issue again. Thanks!

**Comment by byronyi** [comment]
We have observed similar issues with TF 2.1-rc0. Cannot share source code, but hopefully I could get a minimal reproducible case.

```
WARNING:tensorflow:AutoGraph could not transform <bound method TransformerDecoder._transform of <bytedseq.decoders.transformer_decoder.TransformerDecoder object at 0x7f727cc9e390>> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: unexpected indent (<unknown>, line 31)
W1219 06:50:06.045258 140133751701632 ag_logging.py:146] AutoGraph could not transform <bound method TransformerDecoder._transform of <bytedseq.decoders.transformer_decoder.TransformerDecoder object at 0x7f727cc9e390>> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: unexpected indent (<unknown>, line 31)
```

Gently ping @goldiegadde; anything looks familiar?

**Comment by byronyi** [comment]
Gently ping @mdanatg; does this look familiar to you?

**Comment by mdanatg** [comment]
These unexpected indent errors are new to me, but it looks like a regression. Setting `AUTOGRAPH_VERBOSITY=10` should have given you a whole lot of extra logging. Since autograph outputs to the abseil logger, try adding this somewhere early in your code:

```
from absl import logging
logging.set_verbosity(logging.INFO)
```

In the mean time, could you visually inspect the source code of one of the functions that it complains about? Like e.g. `TransformerDecoder._transform` - does it contain any thing out of the ordinary, like mixing tabs with spaces, backslash `\` continuations or anything like that?

As a side note, to get a minimal repro, could you also try the following command: `tf.autograph.to_code(TransformerDecoder._transform)`. It should allow us to test things outside of your main program, even with unbound variables or missing arguments.

**Comment by mjlbach** [comment]
Also having this problem on a fairly innocuous piece of code:
```

def reverse_edges(edges):
    """
    Reverse edges to point from left_node -> right_node to
    right_node -> left_node

    :param edges: dict(('layer', 'idx')), edges

    :return reversed_edges: dict(('layer', idx')), reversed edges
    """
    llayer = edges['layer'][0, 0]
    rlayer = edges['layer'][0, 1]
    reversed_edges = {
            'layer': tf.constant([[rlayer, llayer]]),
            'idx': reverse_edge_idxs(edges['idx']),
            }
    return reversed_edges

def key_to_layers(key):
    """
    Decodes a key into layer numbers.

    :param key: str, node or edge key

    :return llayer: str, left or sender layer number
    :return rlayer: str, right or receiver layer number
    """
    if '_to_' in key:
        llayer, _, rlayer = key.split('_')
    else:
        llayer = rlayer = key.split('_')[-1]
    return llayer, rlayer
        # Reverse edges do not have to be loaded but can be constructed
          if not load_reverse_edges:
            for k in list(edges.keys()):
                llayer, rlayer = key_to_layers(k)
                if llayer != rlayer:
                    edges[connect_key(rlayer, llayer)] = gbuild.reverse_edges(edges[k])
                    edges.pop(k)
```
where edges is a dictionary containing tensors as values. Not a complete example, but maybe might help isolating the issue?

**Comment by mdanatg** [comment]
@mjlbach I tried your code by passing them directly to autograph, e.g. `print(tf.autograph.to_code(build_edges))`. I'm not seeing any errors in Python 3.7; I'll try with 3.5 shortly.

If you have the chance, could you try the same test on the function mentioned in the warning message?

**Comment by mjlbach** [comment]
Ok, the two errors are:
```
WARNING:tensorflow:AutoGraph could not transform <bound method ParticleGraphConstructorFromRecords.construct_hierarchy_from_records of <boxphysics.model.fromrecords.BoxGraphConstructorFromRecords object at 0x7fbbf055fd68>> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: unexpected indent (<unknown>, line 91)
WARNING:tensorflow:AutoGraph could not transform <bound method ParticleGraphConstructorFromRecords.remove_invalid_parts_by_nodes of <boxphysics.model.fromrecords.BoxGraphConstructorFromRecords object at 0x7fbbf055fd68>> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: expected an indented block (<unknown>, line 27)
````

If I try to run autograph_to_code on the first one, I get:
```
*** IndentationError: unexpected indent
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/api.py", line 831, in to_code
    experimental_optional_features=experimental_optional_features))
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/api.py", line 658, in to_graph
    return conversion.convert(entity, program_ctx)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/conversion.py", line 359, in convert
    free_nonglobal_var_names)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/conversion.py", line 274, in _convert_with_cache
    entity, program_ctx)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/conversion.py", line 508, in convert_entity_to_ast
    nodes, name, entity_info = convert_func_to_ast(o, program_ctx)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/conversion.py", line 668, in convert_func_to_ast
    node, source = parser.parse_entity(f, future_features=future_features)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/pyct/parser.py", line 209, in parse_entity
    return _attempt_to_parse_normal_source(source, future_features)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/pyct/parser.py", line 113, in _attempt_to_parse_normal_source
    return parse(source, preamble_len=len(future_features)), source
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/pyct/parser.py", line 226, in parse
    module_node = gast.parse(src)
  File "/usr/local/lib/python3.6/dist-packages/gast/gast.py", line 295, in parse
    return ast_to_gast(_ast.parse(*args, **kwargs))
  File "/usr/lib/python3.6/ast.py", line 35, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
  File "<unknown>", line 91
    if not load_reverse_edges:
```

On the second:
```
*** IndentationError: expected an indented block
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/api.py", line 831, in to_code
    experimental_optional_features=experimental_optional_features))
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/api.py", line 658, in to_graph
    return conversion.convert(entity, program_ctx)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/conversion.py", line 359, in convert
    free_nonglobal_var_names)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/conversion.py", line 274, in _convert_with_cache
    entity, program_ctx)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/conversion.py", line 508, in convert_entity_to_ast
    nodes, name, entity_info = convert_func_to_ast(o, program_ctx)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/conversion.py", line 668, in convert_func_to_ast
    node, source = parser.parse_entity(f, future_features=future_features)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/pyct/parser.py", line 209, in parse_entity
    return _attempt_to_parse_normal_source(source, future_features)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/pyct/parser.py", line 113, in _attempt_to_parse_normal_source
    return parse(source, preamble_len=len(future_features)), source
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/pyct/parser.py", line 226, in parse
    module_node = gast.parse(src)
  File "/usr/local/lib/python3.6/dist-packages/gast/gast.py", line 295, in parse
    return ast_to_gast(_ast.parse(*args, **kwargs))
  File "/usr/lib/python3.6/ast.py", line 35, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
  File "<unknown>", line 27
    llayer, rlayer = key_to_layers(k)
```

**Comment by mdanatg** [comment]
Thank you, and sorry for the delay. It looks like the warning is caused by the same bug as #35765. Removing the backslash continuations worked in my tests. Until that's fixed, any of these workarounds should silence the warning:
 * removing the backslash continuations
 * decorating the function with @tf.autograph.experimental.do_not_convert, since it doesn't contain data-dependent control flow

Note that you can use parentheses to break expressions on multiple lines, as alternative to backslash continuations.

**Comment by mjlbach** [comment]
> Thank you, and sorry for the delay. It looks like the warning is caused by the same bug as #35765. Removing the backslash continuations worked in my tests. Until that's fixed, any of these workarounds should silence the warning:
> 
>     * removing the backslash continuations
> 
>     * decorating the function with @tf.autograph.experimental.do_not_convert, since it doesn't contain data-dependent control flow
> 
> 
> Note that you can use parentheses to break expressions on multiple lines, as alternative to backslash continuations.

Thank you!

**Comment by mdanatg** [comment]
Backslash continuations should now be properly supported; the fix is in tf-nightly and will be available in TF 2.2.

**Comment by ismael-elatifi** [comment]
I still have this issue with TF2.1 and you can find my workaround [here](https://github.com/tensorflow/tensorflow/issues/31373#issuecomment-573663963)
